src/index.ts

```diff
@@ -1,119 +1,147 @@
-
-  public constructor(
-    reconnect: boolean,
-    reconnectEventInit?: CloseEventInit,
-    closeEventInit?: CloseEventInit,
-  ) {
-    super('open', reconnectEventInit);
-
-    this.isReconnect = reconnect;
-    if (closeEventInit) {
-      this.code = closeEventInit.code;
-      this.wasClean = closeEventInit.wasClean;
-      this.reason = closeEventInit.reason;
-    }
-  }
+export interface RwsOptions extends Partial<Options> {
+  /** Should create a `WebSocket` on construction? */
+  automaticOpen?: boolean;
 }
 
+export type WebSocketSendData = Parameters<WebSocket['send']>[0];
+
 /** `EventEmitter` event name to event type map. */
-interface ReconnectingWebSocketEventMap extends Pick<WebSocketEventMap, 'close' | 'message'> {
+export interface ReconnectingWebSocketEventMap extends WebSocketEventMap {
   connecting: ReconnectEvent;
   error: ReconnectEvent;
   open: ReconnectEvent;
 }
 
 // Strongly type `EventEmitter` methods
 interface ReconnectingWebSocket {
+  addListener<K extends keyof ReconnectingWebSocketEventMap>(
+    type: K,
+    listener: (this: this, event: ReconnectingWebSocketEventMap[K]) => void,
+  ): this;
   emit<K extends keyof ReconnectingWebSocketEventMap>(
     type: K,
     event: ReconnectingWebSocketEventMap[K],
   ): boolean;
-  on<K extends keyof ReconnectingWebSocketEventMap>(
+  removeListener<K extends keyof ReconnectingWebSocketEventMap>(
     type: K,
-    listener: (event: ReconnectingWebSocketEventMap[K]) => void,
+    listener: (this: this, event: ReconnectingWebSocketEventMap[K]) => void,
   ): this;
 }
 
 class ReconnectingWebSocket extends EventEmitter implements WebSocket {
-  private forcedClose = false;
-  private reconnectAttempts = 0;
-  private timedOut = false;
-
-  private protocols?: string | string[];
-  private timeout?: number;
-  private ws?: WebSocket;
-
-  public readonly CONNECTING = WebSocket.CONNECTING;
-  public readonly OPEN = WebSocket.OPEN;
   public readonly CLOSING = WebSocket.CLOSING;
   public readonly CLOSED = WebSocket.CLOSED;
+  public readonly CONNECTING = WebSocket.CONNECTING;
+  public readonly OPEN = WebSocket.OPEN;
+
   public readonly DEFAULT_CODE = 1000;
 
   public options: Options;
-  public readyState = WebSocket.CONNECTING;
   public url: string;
 
-  public onconnecting = (event: ReconnectEvent): ReconnectEvent => event;
-  /** Passed `ReconnectEvent` in reality. */
-  public onopen = (event: Event): Event => event;
-  public onclose = (event: CloseEvent): CloseEvent => event;
-  /** Passed `ReconnectEvent` in reality. */
-  public onerror = (event: Event): Event => event;
-  public onmessage = (event: MessageEvent): MessageEvent => event;
+  /**
+   * @deprecated Listener signature is simplified in order to correctly implement `WebSocket`.
+   * Use `addEventListener('close', listener)`.
+   */
+  public onclose: WebSocket['onclose'] = null;
+  public onconnecting:
+    | ((this: this, event: ReconnectingWebSocketEventMap['connecting']) => void)
```

Comment:
this onclose etc methods should still work, as they are valid syntax for interacting with websockets.

Reply:
Yes, and the property methods do still work!
Maybe the `@deprecated` tag is not suitable, but I did want to note that TS clients should prefer the `addListener` method for the correct listener typing.